### ISSUE TITLE
The realloc() in curl_fetch_memory has performance issues on Linux 

### ISSUE THREAD

**Issue by apoliakov** [jeroen/curl#189]
Hello and thanks for your work!

We have a bit of a complex stack. We use `curl_fetch_memory` to make R talk to a database. The exact call is here: https://github.com/Paradigm4/SciDBR/blob/cb9ca02e80e94af2d5f91de8e76c7a3025695715/R/internal.R#L101

Long story short, we found that when repeatedly downloading large result sets (~ 1.5GB or more) the performance is unpredictable. We tracked the variability down to `curl_fetch_memory`. One call may take 2 seconds. The next call may take as much as 200 seconds. Then it may go back down to 2 or 10. Then 50 seconds. And so on. Our "internal ticket" for that is here: https://github.com/Paradigm4/SciDBR/issues/198

Running `perf top` suggested that during those "slow times" the time was spent in the kernel, in the allocator. We then read about your previous "realloc is slow on windows" issue. So we tried essentially removing the `#ifdef` to make it allocate in powers-of-two regardless of the OS. That worked much better for us. The change is trivial, here: 
https://github.com/Paradigm4/curl/commit/a1fb6def14d4dff215f182f3b34636c69f0c4183

So, for example, to fetch a ~1.5-2GB payload - several times - before the fix our successive times, for each fetch, in seconds are:
```
  4.7  
135.0
  6.7
 27.9
  6.7
 64.3
 64.4
 64.4
365.6
 16.8  
```

And after the fix (applied to curl v3.3):
```
 3.5
10.5
 5.8
 6.4
 3.1 
 3.9
 7.9 
 3.4
 2.9
 3.3
 3.1 
 3.5
```

So this change looks a lot better, from our vantage point. Unfortunately the repro is a little annoying and unpredictable. The numbers probably depend on a lot of settings - network, server config, linux kernel version, what other allocations R is making and so on. But we reproduced the behavior on Ubuntu 14 and CentOS 7 - and this change appears to make it much better.

Please let us know if we missed something.

Also - how stable is Master and when might you make a new release?

**Comment by jeroen** [comment]
I guess that sounds resonable, I'll do some small experiments myself. I thought the memory allocator in Linux was pretty well optimized, but perhaps it's better to over-allocate.

> Also - how stable is Master and when might you make a new release?

I'm always working in master so better use the CRAN version. New release probably somewhere in the summer or fall.

**Comment by jeroen** [comment]
Fix in curl 4.0 on CRAN now.
